internal/provider/kubernetes/gateway.go

```diff
@@ -97,28 +103,39 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Req
 	if acceptedClass == nil {
 		r.log.Info("No accepted gatewayclass found for gateway", "namespace", request.Namespace,
 			"name", request.Name)
-		// TODO: Delete gateway from the IR.
-		// xref: https://github.com/envoyproxy/gateway/issues/38
+		for namespacedName := range r.resourceTable.Gateways.LoadAll() {
+			r.resourceTable.Gateways.Delete(namespacedName)
+		}
 		return reconcile.Result{}, nil
 	}
 
 	var allGateways *gwapiv1b1.GatewayList
 	if err := r.client.List(ctx, allGateways); err != nil {
 		return reconcile.Result{}, fmt.Errorf("error listing gateways")
 	}
+
 	// Get all the Gateways for the Accepted=true GatewayClass.
 	acceptedGateways := gatewaysOfClass(acceptedClass, allGateways)
 	if len(acceptedGateways) == 0 {
 		r.log.Info("No gateways found for accepted gatewayclass")
-		// TODO: Delete gateway from the IR.
-		// xref: https://github.com/envoyproxy/gateway/issues/34
-		// xref: https://github.com/envoyproxy/gateway/issues/38
-		return reconcile.Result{}, nil
+	}
+	found := false
+	for i := range acceptedGateways {
+		key := types.NamespacedName{
+			Name:      acceptedGateways[i].GetName(),
+			Namespace: acceptedGateways[i].GetNamespace(),
+		}
+		r.resourceTable.Gateways.Store(key, &acceptedGateways[i])
+		if key == request.NamespacedName {
```

Comment:
wouldn't it be easier to reset `r.resourceTable.Gateways` and fill it up based on the contents of `acceptedGateways` ?

Reply:
Yes, let's fix in a follow-up PR.

Reply:
Agree with ^^, that should also better handle the case where the accepted GatewayClass changes (i.e. there used to be gc-1 and gc-2, with gc-1 being accepted, and then gc-1 gets deleted, now we should be processing gc-2) and we now need to remove all the Gateways for the old GatewayClass, and add the new ones.